src/autobattle.cpp

```diff
@@ -0,0 +1,339 @@
+#include "autobattle.h"
+#include "game_actor.h"
+#include "game_enemy.h"
+#include "game_enemyparty.h"
+#include "game_party.h"
+#include "game_battlealgorithm.h"
+#include "game_battle.h"
+#include "algo.h"
+#include "player.h"
+#include "output.h"
+#include <lcf/reader_util.h>
+#include <lcf/data.h>
+
+namespace AutoBattle {
+
+#ifdef EP_DEBUG_AUTOBATTLE
+template <typename... Args>
+static void DebugLog(const char* fmt, Args&&... args) {
+	Output::Debug(fmt, std::forward<Args>(args)...);
+}
+#else
+template <typename... Args>
+static void DebugLog(const char*, Args&&...) {}
+#endif
+
+constexpr decltype(RpgRtCompat::name) RpgRtCompat::name;
+constexpr decltype(AttackOnly::name) AttackOnly::name;
+constexpr decltype(RpgRtImproved::name) RpgRtImproved::name;
+
+std::unique_ptr<AlgorithmBase> CreateAlgorithm(StringView name) {
+	if (name == RpgRtImproved::name) {
+		return std::make_unique<RpgRtImproved>();
+	}
+	if (name == AttackOnly::name) {
+		return std::make_unique<AttackOnly>();
+	}
+	return std::make_unique<RpgRtCompat>();
+}
+
+void AlgorithmBase::SetAutoBattleAction(Game_Actor& source) {
+	vSetAutoBattleAction(source);
```

Comment:
Is this some convention to use ``v``?
I have a similiar "call forwarder to virtual" in my unfinished Filesystem code.

There I use ``XXXXX`` and for virtual ``XXXXImpl``.

Opinion?

Reply:
This is the non-virtual idiom NVI
http://www.gotw.ca/publications/mill18.htm

If I were a more disciplined NVI folllower, I would do this for all virtual functions. I like the `v` prefix, but I don't really care that much what naming scheme is used.

This is also a good way to fix the clang-tidy warnings on `GetBaseAtk()` etc..